File: internal/widgets/sysinfo/sysinfo.go
package sysinfo

import (
	"bytes"
	_ "embed"
	"encoding/json"
	"fmt"
	"html/template"

	"github.com/andresbott/dashi/internal/sysinfo"
	"github.com/andresbott/dashi/internal/widgets"
)

//go:embed sysinfo.html
var sysinfoHTML string

var tmpl = template.Must(template.New("sysinfo").Parse(sysinfoHTML))

type sysinfoConfig struct {
	ShowMemory bool     `json:"showMemory"`
	ShowUptime bool     `json:"showUptime"`
	Disks      []string `json:"disks"`
}

type diskData struct {
	Mountpoint string
	UsedPct    float64
	TotalHuman string
	FreeHuman  string
}

type sysinfoData struct {
	Disks         []diskData
	ShowMemory    bool
	MemUsedPct    float64
	MemTotalHuman string
	MemUsedHuman  string
	ShowUptime    bool
	UptimeHuman   string
}

func humanBytes(b uint64) string {
	const (
		kb = 1024
		mb = 1024 * kb
		gb = 1024 * mb
		tb = 1024 * gb
	)
	switch {
	case b >= tb:
		return fmt.Sprintf("%.1f TB", float64(b)/float64(tb))
	case b >= gb:
		return fmt.Sprintf("%.1f GB", float64(b)/float64(gb))
	case b >= mb:
		return fmt.Sprintf("%.1f MB", float64(b)/float64(mb))
	case b >= kb:
		return fmt.Sprintf("%.1f KB", float64(b)/float64(kb))
	default:
		return fmt.Sprintf("%d B", b)
	}
}

func humanUptime(seconds uint64) string {
	days := seconds / 86400
	hours := (seconds % 86400) / 3600
	minutes := (seconds % 3600) / 60
	if days > 0 {
		return fmt.Sprintf("%dd %dh %dm", days, hours, minutes)
	}
	if hours > 0 {
		return fmt.Sprintf("%dh %dm", hours, minutes)
	}
	return fmt.Sprintf("%dm", minutes)
}

func NewStaticRenderer() func(json.RawMessage, widgets.RenderContext) (template.HTML, error) {
	return func(config json.RawMessage, _ widgets.RenderContext) (template.HTML, error) {
		var cfg sysinfoConfig
		if len(config) > 0 {
			if err := json.Unmarshal(config, &cfg); err != nil {
				return "", fmt.Errorf("sysinfo config: %w", err)
			}
		}

		info, err := sysinfo.Get()
		if err != nil {
			return "", fmt.Errorf("sysinfo get: %w", err)
		}

		enabledDisks := make(map[string]bool)
		for _, mp := range cfg.Disks {
			enabledDisks[mp] = true
		}

		data := sysinfoData{
			ShowMemory: cfg.ShowMemory,
			ShowUptime: cfg.ShowUptime,
		}

		for _, d := range info.Disks {
			if !enabledDisks[d.Mountpoint] {
				continue
			}
			data.Disks = append(data.Disks, diskData{
				Mountpoint: d.Mountpoint,
				UsedPct:    d.UsedPct,
				TotalHuman: humanBytes(d.Total),
				FreeHuman:  humanBytes(d.Free),
			})
		}

		if cfg.ShowMemory {
			data.MemUsedPct = info.MemUsedPct
			data.MemTotalHuman = humanBytes(info.MemTotal)
			data.MemUsedHuman = humanBytes(info.MemUsed)
		}

		if cfg.ShowUptime {
			data.UptimeHuman = humanUptime(info.UptimeSeconds)
		}

		var buf bytes.Buffer
		if err := tmpl.Execute(&buf, data); err != nil {
			return "", fmt.Errorf("sysinfo render: %w", err)
		}
		return template.HTML(buf.String()), nil
	}
}

File: internal/widgets/sysinfo/sysinfo.js
import { poll } from '/_dashi/assets/dashi.js'

// Mirrors humanBytes in image.go. Both sides must agree: Go prefills the
// value and JS overwrites it 30 seconds later.
export function humanBytes(b) {
    const kb = 1024, mb = 1024 * kb, gb = 1024 * mb, tb = 1024 * gb
    if (b >= tb) return `${(b / tb).toFixed(1)} TB`
    if (b >= gb) return `${(b / gb).toFixed(1)} GB`
    if (b >= mb) return `${(b / mb).toFixed(1)} MB`
    if (b >= kb) return `${(b / kb).toFixed(1)} KB`
    return `${b} B`
}

// Mirrors humanUptime in image.go.
export function humanUptime(seconds) {
    const days = Math.floor(seconds / 86400)
    const hours = Math.floor((seconds % 86400) / 3600)
    const minutes = Math.floor((seconds % 3600) / 60)
    if (days > 0) return `${days}d ${hours}h ${minutes}m`
    if (hours > 0) return `${hours}h ${minutes}m`
    return `${minutes}m`
}

poll('.dashi-sysinfo', {
    url: '/api/v0/widgets/sysinfo',
    every: 30,
    apply: (el, info, set) => {
        for (const item of el.querySelectorAll('[data-disk]')) {
            const disk = (info.disks || []).find(d => d.mountpoint === item.dataset.disk)
            if (!disk) continue
            const fill = item.querySelector('.dashi-sysinfo__bar-fill')
            if (fill) fill.style.width = `${Math.round(disk.usedPct)}%`
            const value = item.querySelector('.dashi-sysinfo__value')
            if (value) value.textContent = `${humanBytes(disk.free)} free / ${humanBytes(disk.total)}`
        }
        set('[data-mem] .dashi-sysinfo__bar-fill', { style: { width: `${Math.round(info.memUsedPct)}%` } })
        set('[data-mem] .dashi-sysinfo__value', `${humanBytes(info.memUsed)} / ${humanBytes(info.memTotal)}`)
        set('[data-uptime] .dashi-sysinfo__value', humanUptime(info.uptimeSeconds))
    },
})
